File: Top Interview 150/C++ Code/Stack/150-evaluate-reverse-polish-notation.cpp
// Time:o(n), space:O(n)
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        int ans=0;
        stack<int>st;
        for(int i=0;i<tokens.size();i++){
            if(tokens[i]=="+" || tokens[i]=="-" || tokens[i]=="*" || tokens[i]=="/"){
                int a=st.top();st.pop();
                int b=st.top(); st.pop();
                if(tokens[i]=="+") st.push(b+a);
                else if(tokens[i]=="-") st.push(b-a);
                else if(tokens[i]=="*") st.push(b*a);
                else if(tokens[i]=="/") st.push(b/a);
                
            }
            else st.push(stoi(tokens[i]));
        }    
        return st.top();

    }
};

File: Top Interview 150/Java Code/Stack/150-evaluate-reverse-polish-notation.java
class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> st = new Stack<>();
        for (int i = 0; i < tokens.length; i++) {
            if (tokens[i].equals("+") || tokens[i].equals("-") || tokens[i].equals("*") || tokens[i].equals("/")) {
                int a = st.pop();
                int b = st.pop();
                if (tokens[i].equals("+")) st.push(b + a);
                else if (tokens[i].equals("-")) st.push(b - a);
                else if (tokens[i].equals("*")) st.push(b * a);
                else if (tokens[i].equals("/")) st.push(b / a);
            } else {
                st.push(Integer.parseInt(tokens[i]));
            }
        }
        return st.peek();
    }
}
